keydctl.8
.TH KEYD 8
.SH NAME
keydctl \- control an onak keyd instance
.SH SYNOPSIS
.PP
.B keydctl
[
.B options
]
.B command
.SH DESCRIPTION
.PP
keydctl is a command line client for interacting with a backend keyd
daemon. It's intended to perform functions that are specifically related
to keyd rather than being generic keyserver functions. See
.BR onak(1)
for details about how to perform key related operations.
.SS "Options"
.TP
\fB\-c \fIFILE\fR\fR
Use \fIFILE\fR as the config file instead of the default.
.TP
\fB\-h\fR
Display help text.
.SS "Commands"
.TP
.B check
Query if keyd is running and accepting commands. Returns 0 if it is, 1
otherwise. Outputs nothing to stdout/stderr.
.TP
.B quit
Request that keyd exits cleanly
.TP
.B status
Display the status of a running keyd. Currently the protocol version in use,
when keyd was started, the number of client connections seen and a breakdown
of the commands seen.
.SH FILES
.br
.nf
.\" set tabstop to longest possible filename, plus a wee bit
.ta \w'/usr/lib/perl/getopts.pl   'u
\fI/etc/onak.conf\fR	default configuration file
.SH NOTES
This man page could probably do with some more details.
.SH "SEE ALSO"
.BR onak (1)
.BR keyd (8)
.SH AUTHOR
onak was written by Jonathan McDowell <noodles@earth.li>. It can be found at
http://www.earth.li/projectpurple/progs/onak.html


HISTORY
0.0.1 - 16th May 2002.

* First release.
* Merges gpgstats 0.0.2 (never released).

0.0.2 - 28th May 2002.

* Added support for subpacket type 0x83 (critical key expiration)
* Fixed bug in parsing of one byte new format packet length.
* Added support for 4 byte old format packet lengths.
* Changed various error statements to output to stderr instead of stdout.
* Fixed bug in deleting keys in Postgres backend.
* Moved code to merge a list of keys to merge.c and changed add.c to use it
  rather than just replacing existing keys.
* Added comment & trust packets as known but ignored types.
* Added storage of uids to Postgres backend; this speeds up verbose indexing
  and will also allow searching on UIDs.
* Fixed bug in armor handling (didn't like 2 newlines before the armor header).
* Made dearmor gracefully handle non armored input.
* Added support for incoming incremental update mails.
* Added searching on uids to lookup & keydb_pg.
* Changed Postgres backend to use 64 bit keyids instead of 32 bit.
* Made dearmor ignore CRs when looking for 2 blank lines.

0.0.3 - 2nd July 2002

* Added index on keyid for key table in Postgres backend.
* Twiddled transaction support in an attempt to speed up mass key adding.
* Changed putchar_func to take a character count rather than doing one char at
  a time (massive speed up in Postgres case).
* Introduced onak binary for general keyserver operations.
* Changed all version number references to use VERSION macro.
* Made lldel free the unused list structure. (pointed out by Simon Huggins)
* Made llfind assert that the cmp function is non NULL. (Simon Huggins again)
* Fixed gpgwww; after the move to 64bit keyids internally it was trying to
  compare the 32bit user supplied keyid to the retrieved 64bit one and never
  finding paths.
* Various tidying up I've forgotten.
* Fixed bug with removing signed packets while merging.
* Fixed bug with potentially adding an already existing uid to a key when
  merging.
* Fixed stupid typo bug in merging keys.

0.0.4 - 10th November 2002

* Made keydb_file backend compile again.
* Changed merging to compare signature keyids rather than packet contents
  so we don't add the same signature multiple times.
* Changed keydb_pg backend over to using PQescapeString to escape SQL data.
* Added list of keyids in a path to gpgwww output (suitable for C&P into
  a gnupg command line for example).
* Pulled out HTML start/end code to getcgi.c and added a DOCTYPE and charset
  of utf8.
* Did some work on getting onak-mail.pl to output incrementals as well as
  receiving them.
* Some cleanup in getting signature keyids.
* Made sure we freed more things after we've finished with them.
* Changed maxpath to display the path found.
* Added use of onak_sigs to keydb_pg to speed up retrieval of key sigs. Should
  speed up the pathfinder a lot.
* Added llfree for freeing up linked lists.
* Wrote DB3 backend; although much work is still needed it vastly outperforms
  the PostgreSQL backend and should hopefully not suffer from the problems of
  the DB2 implementation.

0.1.0 - 13th November 2002

* Code cleanup; pull out common character functions.
* Make DB3 backend support proper searching.
* Added config file.
* Fixed ASCII armor bug (we'd output NULLs in the headers/footer).
* Added bidirectional syncing (previously we could receive but wouldn't send).

0.1.1 - 2nd December 2002

* Fixed utf8 content type - should be UTF-8. (Thanks to Simon Huggins)
* Made getfullkeyid not assert if the key isn't found. (Thanks to Simon 
  Huggins for reporting this.)
* Fixed onak-mail.pl to check the stdout of onak before stderr to prevent
  blocking when reading the output update.
* Used ccmalloc to find various memory leaks and fixed them.
* Added display of key subkeys.
* Tightened up table creation SQL a bit.
* Fixed bug with merging keys that have no new content (we'd sometimes stop
  processing the incoming stream of keys).
* Updated README to be a bit more useful.
* Code cleanup; removed circular dependencies.
* Added sixdegrees.

0.1.2 - 15th February 2003

* Added fingerprint calculation/display.
* Cleanup sixdegrees binary/object file on make clean.
* Introduced transaction support to DB3 backend, along with deadlock detection.
* Added keydb dumping ability.
* Added logging infrastructure to help more easily track down problems.

0.2.0 - 8th June 2003

* Output multiple paths in gpgwww (thanks to Simon Huggins).
* Allow a keyid on the command line for sixdegrees (Simon Huggins).
* Make db2 backend check for a num_keydb file to know how many db files to use.
* Add 0x to generated URLs as pks needs these.
* Add "Find Reverse Path" link to gpgwww output.
* Checkpoint the db3 database upon clean exit.
* Fix bug with logging where the month was one less than it should have been.
* Fall back to stderr if we can't open the logfile.
* Move dependancy list from Makefile to separate file.
* Checkpoint the DB on clean shutdown.
* First cut at MRHKP support.
* Clean up various compile warnings under gcc 3.3
* Log a critical error when we can't handle a critical subpacket rather than
  asserting.
* Make the Postgres backend compile again.
* First attempt at supporting revoked keys.

0.2.1 - 11th October 2003

* Added support for multiple backend DB3 key files.
* Fixed DB3 database dumping to be outside a transaction (otherwise the
  transaction is too big and we run out of memory).
* Change over onak-mail.pl to use the config file more.
* Fix bug where we always read one byte from stdin in rather than allowing
  zero.
* Tightened up error checking in a couple of places to allow proper DB cleanup.
* Changed a printf error message over to using logthing.
* Allow read_openpgp_stream to append to an existing packet list.
* Tidy up various character functions into charfuncs.c
* Add splitkeys for spliting up keyrings.
* Various code cleanups to help reduce warnings under C99.
* Add (verbose) index support to onak-mail.pl

0.3.0 - 14th September 2004


keydctl.c
/*
 * keydctl.c - A simple program to control a running keyd instance
 *
 * Copyright 2011 Jonathan McDowell <noodles@earth.li>
 */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>

#include "keyd.h"
#include "onak-conf.h"
#include "version.h"

/* HACK: We need to stop onak-conf.o requiring this. */
void *DBFUNCS = NULL;

static int keyd_fd = -1;
static int verbose = 0;

static int keyd_do_command(enum keyd_ops cmd, void *buf, size_t len)
{
	uint32_t tmp;

	if (keyd_fd < 0) {
		return -1;
	}

	tmp = cmd;
	if (write(keyd_fd, &tmp, sizeof(tmp)) != sizeof(tmp)) {
		if (verbose >= 0) {
			fprintf(stderr,
				"Couldn't write keyd command %d: %s (%d)\n",
				cmd, strerror(errno), errno);
		}
		exit(EXIT_FAILURE);
	} else if (read(keyd_fd, &tmp, sizeof(tmp)) != sizeof(tmp)) {
		if (verbose >= 0) {
			fprintf(stderr,
				"Couldn't read keyd command %d reply: "
				"%s (%d)\n",
				cmd, strerror(errno), errno);
			}
		exit(EXIT_FAILURE);
	} else if (tmp != KEYD_REPLY_OK) {
		return -1;
	} else if (buf == NULL) {
		return 0;
	} else if (read(keyd_fd, &tmp, sizeof(tmp)) != sizeof(tmp)) {
		if (verbose >= 0) {
			fprintf(stderr,
				"Couldn't read keyd command %d reply length: "
				"%s (%d)\n",
				cmd, strerror(errno), errno);
		}
		exit(EXIT_FAILURE);
	} else if (tmp > len) {
		/* TODO: Read what we can into buf and skip the rest */
		return -1;
	} else {
		return read(keyd_fd, buf, tmp);
	}
}

static void keyd_connect(void)
{
	struct sockaddr_un sock;
	uint32_t	   reply = KEYD_REPLY_UNKNOWN_CMD;

	keyd_fd = socket(PF_UNIX, SOCK_STREAM, 0);
	if (keyd_fd < 0) {
		if (verbose >= 0) {
			fprintf(stderr,
				"Couldn't open socket: %s (%d)\n",
				strerror(errno),
				errno);
		}
		exit(EXIT_FAILURE);
	}

	sock.sun_family = AF_UNIX;
	snprintf(sock.sun_path, sizeof(sock.sun_path) - 1, "%s/%s",
			config.db_dir,
			KEYD_SOCKET);
	if (connect(keyd_fd, (struct sockaddr *) &sock, sizeof(sock)) < 0) {
		if (verbose >= 0) {
			fprintf(stderr,
				"Couldn't connect to socket %s: %s (%d)\n",
				sock.sun_path,
				strerror(errno),
				errno);
		}
		exit(EXIT_FAILURE);
	}

	keyd_do_command(KEYD_CMD_VERSION, &reply, sizeof(reply));
	if (reply != keyd_version) {
		if (verbose >= 0) {
			fprintf(stderr, "Error! keyd protocol version "
				"mismatch. (us = %d, it = %d)\n",
				keyd_version, reply);
		}
		exit(EXIT_FAILURE);
	}

	return;
}

static void keyd_close(void)
{
	uint32_t cmd = KEYD_CMD_CLOSE;

	if (write(keyd_fd, &cmd, sizeof(cmd)) != sizeof(cmd) && verbose >= 0) {
		fprintf(stderr, "Couldn't send close cmd: %s (%d)\n",
				strerror(errno),
				errno);
	}

	if (shutdown(keyd_fd, SHUT_RDWR) < 0 && verbose >= 0) {
		fprintf(stderr, "Error shutting down socket: %d\n",
				errno);
	}
	if (close(keyd_fd) < 0 && verbose >= 0) {
		fprintf(stderr, "Error closing down socket: %d\n",
				errno);
	}
	keyd_fd = -1;

	return;

}

static void keyd_status(void)
{
	uint32_t reply;
	struct keyd_stats stats;

	keyd_do_command(KEYD_CMD_VERSION, &reply, sizeof(reply));
	printf("Using keyd protocol version %d.\n", reply);

	keyd_do_command(KEYD_CMD_STATS, &stats, sizeof(stats));
	printf("keyd running since %s", ctime(&stats.started));
	printf("%d client connections received\n", stats.connects);

	printf("Command statistics:\n");
	printf("  Version:          %d\n",
		stats.command_stats[KEYD_CMD_VERSION]);
	printf("  Get key:          %d\n", stats.command_stats[KEYD_CMD_GET]);
	printf("  Store key:        %d\n",
		stats.command_stats[KEYD_CMD_STORE]);
	printf("  Delete key:       %d\n",
		stats.command_stats[KEYD_CMD_DELETE]);
	printf("  Search key:       %d\n",
		stats.command_stats[KEYD_CMD_GETTEXT]);
	printf("  Get full keyid:   %d\n",
		stats.command_stats[KEYD_CMD_GETFULLKEYID]);
	printf("  Iterate all keys: %d\n",
		stats.command_stats[KEYD_CMD_KEYITER]);
	printf("  Close:            %d\n",
		stats.command_stats[KEYD_CMD_CLOSE]);
	printf("  Quit:             %d\n", stats.command_stats[KEYD_CMD_QUIT]);
	printf("  Get statistics:   %d\n",
		stats.command_stats[KEYD_CMD_STATS]);
	printf("  Unknown:          %d\n",
		stats.command_stats[KEYD_CMD_UNKNOWN]);

	return;
}

static void usage(void)
{
	puts("keydctl " ONAK_VERSION " - control an onak keyd instance.\n");
	puts("Usage:\n");
	puts("\tonak [options] <command> <parameters>\n");
	puts("\tCommands:\n");
	puts("\tcheck    - check if keyd is running");
	puts("\tquit     - request that keyd cleanly shuts down");
	puts("\tstatus   - display running keyd status");
	exit(EXIT_FAILURE);
}

int main(int argc, char *argv[])
{
	int	 optchar;
	char	*configfile = NULL;

	while ((optchar = getopt(argc, argv, "c:h")) != -1 ) {
		switch (optchar) {
		case 'c':
			configfile = strdup(optarg);
			break;
		case 'h':
		default:
			usage();
			break;
		}
	}

	readconfig(configfile);
	free(configfile);
	configfile = NULL;

	if ((argc - optind) < 1) {
		usage();
	} else if (!strcmp("check", argv[optind])) {
		/* Just do the connect and close quietly */
		verbose = -1;
		keyd_connect();
		keyd_close();
	} else if (!strcmp("status", argv[optind])) {
		keyd_connect();
		keyd_status();
		keyd_close();
	} else if (!strcmp("quit", argv[optind])) {
		keyd_connect();
		keyd_do_command(KEYD_CMD_QUIT, NULL, 0);
		keyd_close();
	} else {
		usage();
	}


	exit(EXIT_SUCCESS);
}
